### ISSUE TITLE
MTP: Qwen3.8-Flash-Next stops mid-code before token limit on mixed-GPU layer split

### ISSUE THREAD

**Issue by GenerelSchwerz** [GenerelSchwerz/llama.cpp#103]
## Symptom

A direct chat-completions coding request with `max_tokens: 30000` and no stop strings ended mid-string with `finish_reason: "stop"` when draft-MTP was enabled. The same request, model, GPU order, MoE-cache budgets, context, and sampling settings reached the full 30,000-token limit when the MTP flags were removed.

| Configuration | Completion tokens | Finish reason | Last visible code |
| --- | ---: | --- | --- |
| MTP depth 3 | 12,427 | `stop` | `item_id="en` (inside a string literal) |
| No MTP | 30,000 | `length` | Still generating code at the requested cap |

The MTP run produced 12 fenced code blocks of the requested 22 before stopping. Both runs used `temperature: 0.7` and no fixed seed, so this A/B strongly implicates the MTP path but does not by itself prove determinism or identify the offending token.

## Environment

- Fork build: `0.4.1-dev`, build 11489, commit `0031a507bfa8921e107c8a804e0973a076b04f6e` (`moe-cache-multi-gpu`)
- Ubuntu 24.04.4, NVIDIA driver 580.173.02, CUDA toolkit 12.8
- CUDA0: V100 16 GB; CUDA1: V100 32 GB; CUDA2: RTX 4090 24 GB; CUDA3/4: V100 16 GB each
- Target: Qwen3.8-Flash-Next Q8_0 GGUF; draft: shared MTP Q8_0 GGUF; one server slot

## Reproduction

Start `llama-server` with these flags (substitute local model paths):

```sh
CUDA_DEVICE_ORDER=PCI_BUS_ID llama-server \
  -m /path/to/Qwen3.8-Flash-Next-Q8_0-00001-of-00006.gguf \
  --mmproj /path/to/mmproj-BF16.gguf --mmproj-device CUDA0 \
  --spec-draft-model /path/to/mtp-Qwen3.8-Flash-Next-shared-Q8_0.gguf \
  --spec-draft-device CUDA2 --spec-draft-ngl all \
  --spec-type draft-mtp --spec-draft-n-max 3 \
  --device CUDA3,CUDA4,CUDA0,CUDA1,CUDA2 --split-mode layer \
  --moe-expert-cache-mib 15850,15850,14950,31350,19500 \
  --load-mode none --lazy-mode off -ngl 99 -c 131072 -np 1 \
  -b 1024 -ub 512 -fa on -ctk q8_0 -ctv q8_0 --fit off \
  --jinja --chat-template-kwargs '{"reasoning_effort":"medium"}' \
  --host 0.0.0.0 --port 8080 -lv 4
```

POST to `/v1/chat/completions` with `stream: false`, `max_tokens: 30000`, `temperature: 0.7`, and this user message:

> Implement a substantial, runnable Python 3.12 text-adventure RPG package from scratch. Deliver at least 20 complete files, each in its own fenced code block with its path as a heading: pyproject.toml, README.md, src/rpg/__init__.py, models.py, world.py, map_generator.py, characters.py, inventory.py, items.py, combat.py, magic.py, quests.py, dialogue.py, merchants.py, crafting.py, persistence.py, commands.py, cli.py, plus tests/test_world.py, tests/test_combat.py, tests/test_quests.py, and tests/test_persistence.py. Include a coherent world, 30 rooms, 20 items, ten enemy types, level progression, turn-based combat, inventory, equipment, quests, NPC dialogue, shops, crafting, save/load, a command-line loop, and complete tests. Write all necessary functions and data; no placeholders, ellipses, omitted sections, or pseudo-code. Ensure imports and APIs agree across files. Do not stop until every file and test is complete. Start with pyproject.toml.

For the control, remove only the five `--spec-*` arguments above. The MTP server reported `eval time = 346084.59 ms / 12427 tokens`; the no-MTP control reported `1305234.29 ms / 30000 tokens`. Both ended with `truncated = 0`; neither server crashed, and health checks remained OK. Two earlier interactive requests also ended mid-response (17,511 and 1,162 generated tokens), but their raw finish reasons were not captured.

## Investigation needed

The raw `stop` result distinguishes this from a client output cap. No stop string was supplied, so an EOG token is the likely server stop path; the specific token was not captured. Check the sampled target token and target logits at the stop, then compare MTP verification/recurrent state against a no-MTP run. Repeat with a fixed seed and varying MTP depth before assigning a root cause.

**Comment by GenerelSchwerz** [comment]
Thanks for the report and the detailed results. I am working on it.
